File: app/schemas/fund_flow.py
"""免费资金流 Pydantic Schemas (方案 C API 契约)。

前端 agent 直接消费这些 schema；JSON 字段名与数据库 ORM 字段保持一致
（snake_case），与 API 契约注释中的 TypeScript 字段顺序一致。
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# 1. 个股资金流
# ---------------------------------------------------------------------------


class IndividualFundFlowBase(BaseModel):
    """个股资金流公共字段。"""

    ts_code: str
    trade_date: date
    main_net_inflow: float | None = None
    main_net_pct: float | None = None
    super_large_net: float | None = None
    super_large_pct: float | None = None
    large_net: float | None = None
    large_pct: float | None = None
    medium_net: float | None = None
    medium_pct: float | None = None
    small_net: float | None = None
    small_pct: float | None = None


class IndividualFundFlowOut(IndividualFundFlowBase):
    """API 返回的个股资金流记录。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    source: str = "akshare"
    fetched_at: datetime | None = None


class IndividualFundFlowListResponse(BaseModel):
    """个股资金流列表响应 (分页)。"""

    items: list[IndividualFundFlowOut]
    total: int
    page: int
    page_size: int


# ---------------------------------------------------------------------------
# 2. 板块资金流
# ---------------------------------------------------------------------------


class SectorFundFlowBase(BaseModel):
    """板块资金流公共字段。"""

    sector_name: str
    sector_type: str = Field(..., description="板块类型: 行业 / 概念 / 地域")
    trade_date: date
    main_net_inflow: float | None = None
    main_net_pct: float | None = None
    super_large_net: float | None = None
    large_net: float | None = None
    leading_stock: str | None = None


class SectorFundFlowOut(SectorFundFlowBase):
    """API 返回的板块资金流记录。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    fetched_at: datetime | None = None


class SectorFundFlowListResponse(BaseModel):
    """板块资金流列表响应。"""

    items: list[SectorFundFlowOut]
    total: int
    page: int
    page_size: int


# ---------------------------------------------------------------------------
# 3. ETF 资金流
# ---------------------------------------------------------------------------


class EtfFundFlowBase(BaseModel):
    """ETF 资金流公共字段。"""

    ts_code: str
    trade_date: date
    price: float | None = None
    net_value: float | None = None
    premium_rate: float | None = None
    shares_outstanding: float | None = None
    shares_change: float | None = None
    turnover: float | None = None
    inferred_net_inflow: float | None = None


class EtfFundFlowOut(EtfFundFlowBase):
    """API 返回的 ETF 资金流记录。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    fetched_at: datetime | None = None


class EtfFundFlowListResponse(BaseModel):
    """ETF 资金流列表响应。"""

    items: list[EtfFundFlowOut]
    total: int
    page: int
    page_size: int


# ---------------------------------------------------------------------------
# 4. 综合资金信号
# ---------------------------------------------------------------------------


class FlowSignalBase(BaseModel):
    """综合资金信号公共字段。"""

    ts_code: str
    trade_date: date
    main_net_inflow: float | None = None
    margin_net_change: float | None = None
    lhb_net_buy: float | None = None
    shareholder_count_change: float | None = None
    ah_premium: float | None = None
    block_trade_net: float | None = None
    composite_score: float | None = None
    score_breakdown: dict[str, Any] | None = None


class FlowSignalOut(FlowSignalBase):
    """API 返回的综合资金信号记录。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    fetched_at: datetime | None = None


class FlowSignalListResponse(BaseModel):
    """综合资金信号列表响应。"""

    items: list[FlowSignalOut]
    total: int
    page: int
    page_size: int


# ---------------------------------------------------------------------------
# 5. 大盘整体资金流 (ak.stock_market_fund_flow)
# ---------------------------------------------------------------------------


class MarketFundFlowOut(BaseModel):
    """大盘整体资金流 (ak.stock_market_fund_flow 一次返回两行)。"""

    trade_date: date = Field(..., description="交易日期 (akshare 上证/深证 同步)")
    sh_main_net_inflow: float | None = None
    sz_main_net_inflow: float | None = None
    sh_main_net_pct: float | None = None
    sz_main_net_pct: float | None = None


class MarketFundFlowListResponse(BaseModel):
    """大盘资金流多日响应。"""

    items: list[MarketFundFlowOut]
    total: int

File: web/src/api/fundFlow.ts
import { useQuery, useQueryClient } from '@tanstack/react-query';
import client from './client';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Sort direction helpers. Backend expects `field` or `-field` (leading
 *  minus = desc). Kept here so call sites stay type-safe. */
export type SortDir = 'asc' | 'desc';

export interface MarketFundFlow {
  trade_date: string;
  sh_main_net_inflow: number;
  sz_main_net_inflow: number;
  sh_main_net_pct: number;
  sz_main_net_pct: number;
  total_main_net_inflow?: number;
  total_main_net_pct?: number;
}

export type FundFlowSource = 'akshare' | 'eastmoney';

export interface IndividualFundFlow {
  ts_code: string;
  name?: string | null;
  trade_date: string;
  main_net_inflow: number;
  main_net_pct: number;
  super_large_net: number;
  super_large_pct: number;
  large_net: number;
  large_pct: number;
  medium_net: number;
  medium_pct: number;
  small_net: number;
  small_pct: number;
  source: FundFlowSource;
}

export type SectorType = '行业' | '概念' | '地域';

export interface SectorFundFlow {
  sector_name: string;
  sector_type: SectorType;
  trade_date: string;
  main_net_inflow: number;
  main_net_pct: number;
  super_large_net: number;
  large_net: number;
  leading_stock?: string | null;
}

export interface EtfFundFlow {
  ts_code: string;
  name?: string | null;
  trade_date: string;
  price: number;
  net_value: number;
  premium_rate: number;
  shares_outstanding: number;
  shares_change: number;
  turnover: number;
  inferred_net_inflow: number;
}

export interface FlowSignal {
  ts_code: string;
  name?: string | null;
  trade_date: string;
  main_net_inflow: number;
  margin_net_change: number;
  lhb_net_buy: number;
  shareholder_count_change: number;
  ah_premium: number;
  block_trade_net: number;
  composite_score: number;
  score_breakdown: Record<string, number>;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Convert a `SortDir` to the backend's `field` / `-field` notation.
 * Backend uses FastAPI's `sort` query param which honors leading minus as
 * descending. Centralizing this avoids `-1` / `desc` / `false` drift across
 * call sites.
 */
export function sortField(field: string, dir: SortDir = 'desc'): string {
  return dir === 'desc' ? `-${field}` : field;
}

// ---------------------------------------------------------------------------
// API client
// ---------------------------------------------------------------------------

export interface IndividualListParams {
  trade_date?: string;
  sort?: string;
  limit?: number;
  market?: string;
}

export interface SectorListParams {
  trade_date?: string;
  sector_type?: SectorType;
  sort?: string;
}

export interface EtfListParams {
  trade_date?: string;
  sort?: string;
  limit?: number;
}

export interface SignalListParams {
  trade_date?: string;
  sort?: string;
  limit?: number;
}

export const fundFlowApi = {
  /**
   * Aggregated exchange-level main-fund net inflow for a given trade date.
   * SH + SZ main flows plus their share-of-turnover percentage.
   */
  market(trade_date?: string): Promise<{ data: MarketFundFlow }> {
    return client.get<MarketFundFlow>('/fund-flow/market', { params: { trade_date } });
  },

  /**
   * Per-stock main-fund flow ranked by `sort`. Defaults to
   * `-main_net_inflow` (top inflow). Optional `market` filter narrows
   * to SH / SZ / STAR / ChiNext / BSE — caller decides which subset to
   * surface (the backend is permissive about the value).
   */
  individualList(params: IndividualListParams = {}): Promise<{ data: IndividualFundFlow[] }> {
    return client.get<IndividualFundFlow[]>('/fund-flow/individual', { params });
  },

  /** Per-stock daily flow history for the trailing N trading days. */
  individualHistory(tsCode: string, days = 60): Promise<{ data: IndividualFundFlow[] }> {
    return client.get<IndividualFundFlow[]>(
      `/fund-flow/individual/${encodeURIComponent(tsCode)}`,
      { params: { days } },
    );
  },

  /**
   * Per-sector flow ranked by `sort`. `sector_type` is a literal string
   * the backend stores as-is ("行业" / "概念" / "地域") so we keep the
   * same casing when round-tripping.
   */
  sectorList(params: SectorListParams = {}): Promise<{ data: SectorFundFlow[] }> {
    return client.get<SectorFundFlow[]>('/fund-flow/sector', { params });
  },

  /** Per-ETF inferred flow (shares-change × price) and premium/discount. */
  etfList(params: EtfListParams = {}): Promise<{ data: EtfFundFlow[] }> {
    return client.get<EtfFundFlow[]>('/fund-flow/etf', { params });
  },

  /**
   * Composite flow signals — cross-source score in `[-100, +100]` plus a
   * breakdown of contributing drivers. Higher = more inflow conviction.
   */
  signalsList(params: SignalListParams = {}): Promise<{ data: FlowSignal[] }> {
    return client.get<FlowSignal[]>('/fund-flow/signals', { params });
  },

  signalsHistory(tsCode: string, days = 30): Promise<{ data: FlowSignal[] }> {
    return client.get<FlowSignal[]>(
      `/fund-flow/signals/${encodeURIComponent(tsCode)}`,
      { params: { days } },
    );
  },
};

// ---------------------------------------------------------------------------
// React Query hooks (used by the page)
// ---------------------------------------------------------------------------

const SHARED_OPTIONS = {
  staleTime: 60_000,
  refetchOnWindowFocus: false,
} as const;

/** Today's main-fund exchange-level flow. Optional `trade_date` for history. */
export function useFundFlowMarket(trade_date?: string) {
  return useQuery({
    queryKey: ['fund-flow', 'market', trade_date ?? 'today'],
    queryFn: () => fundFlowApi.market(trade_date).then((r) => r.data),
    ...SHARED_OPTIONS,
  });
}

/** Per-stock flow list. The page passes its own params object. */
export function useFundFlowIndividual(params: IndividualListParams) {
  return useQuery({
    queryKey: ['fund-flow', 'individual', params],
    queryFn: () => fundFlowApi.individualList(params).then((r) => r.data ?? []),
    ...SHARED_OPTIONS,
  });
}

/** Per-stock daily history (used by the detail placeholder + sparkline). */
export function useFundFlowIndividualHistory(tsCode: string | null, days = 60) {
  return useQuery({
    queryKey: ['fund-flow', 'individual-history', tsCode, days],
    queryFn: () =>
      tsCode
        ? fundFlowApi.individualHistory(tsCode, days).then((r) => r.data ?? [])
        : Promise.resolve([]),
    enabled: !!tsCode,
    staleTime: 5 * 60_000,
    refetchOnWindowFocus: false,
  });
}

export function useFundFlowSector(params: SectorListParams) {
  return useQuery({
    queryKey: ['fund-flow', 'sector', params],
    queryFn: () => fundFlowApi.sectorList(params).then((r) => r.data ?? []),
    ...SHARED_OPTIONS,
  });
}

export function useFundFlowEtf(params: EtfListParams) {
  return useQuery({
    queryKey: ['fund-flow', 'etf', params],
    queryFn: () => fundFlowApi.etfList(params).then((r) => r.data ?? []),
    ...SHARED_OPTIONS,
  });
}

export function useFundFlowSignals(params: SignalListParams) {
  return useQuery({
    queryKey: ['fund-flow', 'signals', params],
    queryFn: () => fundFlowApi.signalsList(params).then((r) => r.data ?? []),
    ...SHARED_OPTIONS,
  });
}

/**
 * Manual refresh helper: invalidate every fund-flow query key so callers
 * can wire the header "刷新" button without re-implementing the cache map.
 */
export function useRefreshFundFlow() {
  const qc = useQueryClient();
  return () => qc.invalidateQueries({ queryKey: ['fund-flow'] });
}